omegaconf/nodes.py

```diff
@@ -33,15 +33,19 @@ def _set_value(self, value: Any, flags: Optional[Dict[str, bool]] = None) -> Non
         if self._get_flag("readonly"):
             raise ReadonlyConfigError("Cannot set value of read-only config node")
 
-        if isinstance(value, str) and get_value_kind(
+        if self._should_validate(value):
+            self._val = self.validate_and_convert(value)
+        else:
+            self._val = value
+
+    def _should_validate(self, value: Any) -> bool:
+        # Validate a value if it is not missing nor an interpolation.
+        return not isinstance(value, str) or get_value_kind(
             value, strict_interpolation_validation=True
-        ) in (
+        ) not in (
             ValueKind.INTERPOLATION,
             ValueKind.MANDATORY_MISSING,
-        ):
-            self._val = value
-        else:
-            self._val = self.validate_and_convert(value)
+        )
```

Comment:
This is pretty hard to parse. can you reorganize it to be easier to understand?

Reply:
Instead of introducing should_validate(), can the new class override _set_value?